### PR TITLE
gh-808 Add support for backend LDAP authentication

### PR DESCRIPTION
**Pull request by ghillert**
* Add documentation
* Add tests
  - Migrate LDAP tests from Spring XD to SCDF
  - Add end-to-end test infrastructure
  - Add Junit rule to bootstrap a local SCDF server
  - Add Junit rule to bootstrap embedded LDAP server

Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/808

### CODE REVIEW

File: spring-cloud-dataflow-server-local/pom.xml [hunk 1]
```diff
@@ -17,6 +17,24 @@
...
+			<groupId>commons-io</groupId>
+			<artifactId>commons-io</artifactId>
+			<version>2.4</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.directory.server</groupId>
+			<artifactId>apacheds-protocol-ldap</artifactId>
+			<version>1.5.5</version>
+			<scope>test</scope>
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-local/pom.xml]: Why are the tests in server-local and not in server-core?

File: spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/local/security/SecurityTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.springframework.security.crypto.codec.Base64;
+
+/**
+ * @author Marius Bogoevici
+ * @author Gunnar Hillert
+ */
+public class SecurityTestUtils {
+	public static String basicAuthorizationHeader(String username, String password) {
+		return "Basic " + new String(Base64.encode((username + ":" + password).getBytes()));
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/local/security/SecurityTestUtils.java]: Should we change the `getBytes()` call?
http://stackoverflow.com/a/7243567/1662330

File: spring-cloud-dataflow-server-local/src/test/resources/org/springframework/cloud/dataflow/server/local/security/testUsers.ldif [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+ou: manager
+member: uid=ben,ou=people,dc=springframework,dc=org
+member: cn=mouse\, jerry,ou=people,dc=springframework,dc=org
+
+dn: cn=submanagers,ou=subgroups,ou=groups,dc=springframework,dc=org
+objectclass: top
+objectclass: groupOfNames
+cn: submanagers
+ou: submanager
+member: uid=ben,ou=people,dc=springframework,dc=org
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-local/src/test/resources/org/springframework/cloud/dataflow/server/local/security/testUsers.ldif]: There are a handful of missing newlines in this PR

File: spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/local/security/LdapServerResource.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	private File workingDir;
+
+	@Override
+	protected void before() throws Throwable {
+		apacheDSContainer = new ApacheDSContainer("dc=springframework,dc=org",
+				"classpath:org/springframework/cloud/dataflow/server/local/security/testUsers.ldif");
+		int ldapPort = SocketUtils.findAvailableTcpPort();
+		apacheDSContainer.setPort(ldapPort);
+		apacheDSContainer.afterPropertiesSet();
+		workingDir = new File(FileUtils.getTempDirectoryPath() + "/" + UUID.randomUUID().toString());
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/local/security/LdapServerResource.java]: Shouldn't we mark as delete-on-exit?

File: spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/local/security/LdapServerResource.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		apacheDSContainer.setWorkingDirectory(workingDir);
+		apacheDSContainer.start();
+		System.setProperty("ldap.port", Integer.toString(ldapPort));
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-local/src/test/java/org/springframework/cloud/dataflow/server/local/security/LdapServerResource.java]: Who uses this?
Should store the previous value (if any) and restore at end of test

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/LdapAuthenticationConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+	public void setGroupRoleAttribute(String groupRoleAttribute) {
+		this.groupRoleAttribute = groupRoleAttribute;
+	}
+
+	@Override
+	public void init(AuthenticationManagerBuilder auth) throws Exception {
+
+		LdapAuthenticationProviderConfigurer<AuthenticationManagerBuilder> ldapConfigurer = auth.ldapAuthentication();
+
+		Assert.hasText(url, "'url' must not be empty");
```
Comment (ericbottard) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/security/LdapAuthenticationConfiguration.java]: Can we move the properties in their own class and/or add jsr303 validation to them?

File: spring-cloud-dataflow-server-local/pom.xml (hunk 1 above)
Comment (ghillert) [inline spring-cloud-dataflow-server-local/pom.xml]: That is no purpose. I am doing full-blown end-to-end testing bootstrapping an entire SCDF server instance. This also allows me to take into account the configuration file from the `spring-cloud-starter-dataflow-server-local` project. 

Ultimately, I am glad that I did that, as that highlighted and issue when enabling security in conjunction with the Boot management REST endpoints.

Lastly, the infrastructure here, might also be used (and refactored) for using in the other implementations such as CF, Yarn etc. Also, being able to bootstrap an entire SCDF instance,  might also be a useful spot to run the Spring Rest Doc tests.

**Comment by ghillert** [comment]
@ericbottard Made a few changes:

* Refactor out Ldap properties into `LdapSecurityProperties`
* Ensure Bean validation for `LdapSecurityProperties`
* Remove reliance on `commons-io`

File: spring-cloud-dataflow-server-local/src/test/resources/org/springframework/cloud/dataflow/server/local/security/testUsers.ldif (hunk 3 above)
Comment (ghillert) [inline spring-cloud-dataflow-server-local/src/test/resources/org/springframework/cloud/dataflow/server/local/security/testUsers.ldif]: Fixed the line feeds.

**Comment by ghillert** [comment]
@ericbottard Did one more commit:

* Fix linefeed issues
* Ensure that basic authorization header is encoded using `ISO_8859_1`

**Comment by ericbottard** [comment]
Rebased, squashed and merged as 07f38ed

**Pull request closed without merging**